pkg/runtime/sequencer/cache.go

```diff
@@ -2,15 +2,198 @@ package sequencer
 
 import (
 	"context"
+	"errors"
 	"mosn.io/layotto/components/sequencer"
+	"mosn.io/pkg/log"
+	"mosn.io/pkg/utils"
+	"sync"
 )
 
+const defaultSize = 1000
+const defaultLimit = 700
+
+// DoubleBuffer is double segment id buffer.
+// There are two buffers in DoubleBuffer: InUseBuffer is in use, BackUpBuffer is a backup buffer.
+// Their default capacity is 1000. When the InUseBuffer usage exceeds 30%, the BackUpBuffer will be initialized.
+// When InUseBuffer is used up, swap them.
+type DoubleBuffer struct {
+	Key          string
+	Size         int
+	InUseBuffer  *Buffer
+	BackUpBuffer *Buffer
+	lock         sync.Mutex
+	Store        sequencer.Store
+}
+
+type Buffer struct {
+	from int64
+	to   int64
+}
+
+func NewDoubleBuffer(key string, store sequencer.Store) *DoubleBuffer {
+
+	d := &DoubleBuffer{
+		Key:   key,
+		Size:  defaultSize,
+		Store: store,
+	}
+
+	return d
+}
+
+//init double buffer
+func (d *DoubleBuffer) init() error {
+
+	buffer, err := d.getNewBuffer()
+	if err != nil {
+		return err
+	}
+
+	d.InUseBuffer = buffer
+
+	return nil
+}
+
+//getId next id
+func (d *DoubleBuffer) getId() (int64, error) {
+
+	if d.InUseBuffer == nil {
+		return 0, errors.New("[DoubleBuffer] Get error: InUseBuffer nil ")
+	}
+
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	//check swap
+	if d.InUseBuffer.from > d.InUseBuffer.to {
+		err := d.swap()
+		if err != nil {
+			return 0, err
+		}
+	}
+	next := d.InUseBuffer.from
+	d.InUseBuffer.from++
+
+	//when InUseBuffer id more than half used, initialize BackUpBuffer.
+	//equal make sure only one thread enter
+	if d.InUseBuffer.to-d.InUseBuffer.from == defaultLimit {
+		utils.GoWithRecover(func() {
+
+			//lock to avoid data race
+			d.lock.Lock()
```

Comment:
the behaviour we expedct is:  if `getId` execeed some threshold, `getNewBuffer` asynchoronus, which won't impact `getId` keep going. but with the huge lock here, async getNewBuffer will **block** getId.

the `DoubleBuffer` is a classic design, but implement it have challenge.

how about do more research and post what you found in this PR.

Reply:
@MoonShining  yeah,  it‘s  a problem. java has `volatile` for visibility.  I don't know what to do on golang. I will do more  research. thanks for your review.